lib/functional.fnl

```diff
@@ -125,6 +125,10 @@
      (set ct (+ ct 1))))
   ct)
 
+(fn apply [f ...]
+  (let [args [...]]
```

Comment:
do you have to assign the name here or can you just return `(f (table.unpack ...))`?

Reply:
This seems to 'double nest' it. E.g. if I call `(apply add [1 2])` in `apply` `args` is `{ { 1, 2} }`

Reply:
I don't think `(f (table.unpack ...))` works.
 
 > if I call (apply add [1 2])
 
 You're giving it a single argument. (apply) usually works with multiple, in your case you'd call it `(apply add 1 2)

Reply:
what's the point of apply, then? Couldn't you just call `(add 1 2)`?. As I understand it (from clojure), `(apply add [1 2]))` should be equivalent to `(add 1 2)`

Reply:
Have you given this one more thought?

Reply:
Hmm... ah... only now I realized what you're talking about... So, in Clojure, apply takes arguments from a sequence, so yes, `(apply 1 2)` and `(apply [1 2])` are identical.

Perhaps here, in Fennel we have to do something like:

```
(fn apply [f ...]
  (let [args [...]]
    (f (table.unpack (table.unpack args)))))
```

so it works with cases like this:

```
(apply some-function 1 2 [3 4])
```
Which would be equivalent to:

```
(some-function 1 2 3 4)
```

Reply:
Or do we have to check the last arg to be of table?

```
(fn apply [f ...]
  (let [args [...]
        last-arg (. args (length args))]
    (if (and (= (type last-arg) "table") (= (length args) 2))
        (f (table.unpack last-arg))
        (f (table.unpack args)))))
 ```
 maan, this is getting a bit crazy. I [checked Clojure's apply,](https://github.com/clojure/clojure/blob/clojure-1.11.1/src/clj/clojure/core.clj#L662) it ain't a one-liner either.

Reply:
I don't think that's how apply should work (in clojure) at least
```
user=> (str 1 2 3 4)
"1234"
user=> (str 1 2 [3 4])
"12[3 4]"
user=> (apply str 1 2 3 4)
java.lang.IllegalArgumentException: Don't know how to create ISeq from: java.lang.Long [at <repl>:3:1]
user=> (apply str [1 2 3 4])
"1234"
```

Reply:
And I figure we should probably work in the same way as clojure

Reply:
Let's put a pin on this one. It definitely requires some more thought. Even[ fennel-cljlib implementation](https://gitlab.com/andreyorst/fennel-cljlib/-/blob/master/src/cljlib.fnl?ref_type=heads#L1152) is not so trivial.

Reply:
oh that's strange. I don't think I have tried to use apply with more than 2 args (a function and a sequence).

I am ok with figuring this out later. We can remove this and merge the rest?

Reply:
hmmm... apply is used though. in `(map)` particularly. I'd say let's merge it as is and then maybe improve it later? I think for general use cases it works. Or do you prefer to get it right instead and not to increase the tech-debt budget?

Reply:
Seems that clojure basically only unpack the last argument, which must be a sequence. This implementation collects all args into a table and then unpacks that, which is also incorrect.

`map` only uses `apply` with two args, so what I think might be best would be to write this apply for 2 args `[fn seq]` and unpack that seq, we can later add support for extra intermediate non-splices args between the function and the last sequence.